File: Client/assets/Script/common/message_handler_manager.ts


export default class MsgHandlerManager
{
    private handlers: Map<number, Function> = null
    constructor() {
        this.handlers = new Map<number, Function>()
    }

    register(msgId: number, handler: Function)
    {
        if (!this.handlers.has(msgId))
        {
            this.handlers.set(msgId, handler)
        }
        else
        {
            console.warn(`message ${msgId} was registered`)
        }
    }

    deregister(msgId: number)
    {
        if (!this.handlers.has(msgId))
        {
            console.warn(`message ${msgId} is never registerd ever`)
            return
        }
        this.handlers.delete(msgId)
    }

    getHandler(msgId: number): Function
    {
        if (this.handlers.has(msgId))
        {
            return this.handlers.get(msgId)
        }
        return null
    }

}

File: Server/src/messages/message_handler_manager.py
import logging
import json
from .in_msgs import InMsgs
from .out_msgs import OutMsgs

class MsgHandlerManager:
    def __init__(self):
        self.msgHandlers = {}

    def register(self, msgId, msgHandler):
        if msgId not in self.msgHandlers:
            self.msgHandlers[msgId] = msgHandler
            return
        logging.warn(f"{msgId} was registered")

    def deregister(self, msgId):
        if msgId not in self.msgHandlers:
            logging.warn(f"{msgId} cannot be deregistered, since no handler is found bind for it")
            return
        del self.msgHandlers[msgId]
        logging.info(f"{msgId} is deregistered")

    def getHandler(self, msgId):
        return self.msgHandlers[msgId] if msgId in self.msgHandlers else None

    def receive(self, addr, data):
        msg = json.loads(data)
        header = msg["header"]
        #body = msg["body"]
        msgId = header["msgId"]
        handler = self.getHandler(msgId)
        if handler is None:
            logging.warn(f"Cannot find handler for message {msgId}")
            return
        handler(addr, msg)
